sky/provision/azure/instance.py

```diff
@@ -45,64 +141,412 @@ def get_azure_sdk_function(client: Any, function_name: str) -> Callable:
+            logger.error(msg)
+            raise RuntimeError(msg)
+
+        resumed_instances = stopped_instances[:to_start_count]
+        resumed_instances.sort(key=lambda x: x.name)
+        resumed_instance_ids = [t.name for t in resumed_instances]
+        logger.debug(
+            f'run_instances: Resuming stopped instances {resumed_instance_ids}.'
+        )
+        start_virtual_machine = _get_azure_sdk_function(
+            compute_client.virtual_machines, 'start')
+        with pool.ThreadPool() as p:
+            p.starmap(
+                start_virtual_machine,
+                [(resource_group, inst.name) for inst in resumed_instances])
+
+    to_start_count -= len(resumed_instance_ids)
+
+    if to_start_count > 0:
+        resource_client = azure.get_client('resource', subscription_id)
+        logger.debug(f'run_instances: Creating {to_start_count} instances.')
+        created_instances = _create_instances(
+            compute_client=compute_client,
+            resource_client=resource_client,
+            cluster_name_on_cloud=cluster_name_on_cloud,
+            resource_group=resource_group,
+            provider_config=provider_config,
+            node_config=config.node_config,
+            tags=tags,
+            count=to_start_count)
+        created_instance_ids = [inst.name for inst in created_instances]
+
+    non_running_instances = list(
+        set(AzureInstanceStatus) - {AzureInstanceStatus.RUNNING})
+    while True:
+        # Wait for all instances to be in running state
+        instances = _filter_instances(compute_client,
+                                      resource_group,
+                                      filters,
+                                      status_filters=non_running_instances,
+                                      included_instances=created_instance_ids +
+                                      resumed_instance_ids)
+        if not instances:
+            break
+        logger.debug(f'run_instances: Waiting for {len(instances)} instances '
+                     'in PENDING status.')
+        time.sleep(_POLL_INTERVAL)
+
+    running_instances = _filter_instances(
+        compute_client,
+        resource_group,
+        filters,
+        status_filters=[AzureInstanceStatus.RUNNING])
+    head_instance_id = _get_head_instance_id(running_instances)
+    instances_to_tag = copy.copy(running_instances)
+    if head_instance_id is None:
+        head_instance_id = _create_instance_tag(instances_to_tag[0])
+        instances_to_tag = instances_to_tag[1:]
+    else:
+        instances_to_tag = [
+            inst for inst in instances_to_tag if inst.name != head_instance_id
+        ]
+
+    if instances_to_tag:
+        # Tag the instances in case the old resumed instances are not correctly
+        # tagged.
+        with pool.ThreadPool() as p:
+            p.starmap(_create_instance_tag,
+                      [(inst, False) for inst in instances_to_tag])
```

Comment:
could we add a comment for the meaning of `False` here? (iiuc it is `is_head=False`?)

Reply:
This comment looks great to me!